### PR TITLE
feat: create donation module in impact dashboard

### PR DESCRIPTION
**Pull request by christian14b**
- Add new "Your donations" module
- Remove the “average donations per $25” stat from “Your lending insights” module

image

### CODE REVIEW

File: src/pages/Portfolio/ImpactDashboard/YourDonations.vue [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+					if (amountOfLoans.value() > 0) {
+						donation.multiply(1 / amountOfLoans.value()).multiply(25);
+					}
+					this.averageDonation = donation.format('$0,0[.]00');
+
+					this.totalDonations = data?.my?.transactions?.totalCount ?? 0;
+					this.currentYearDonations = this.filterAndSumDonations(transactions, currentYear);
+					this.lastYearDonations = this.filterAndSumDonations(transactions, currentYear - 1);
+
+					if (this.totalDonations > this.limit) {
```
Comment (dyersituations) [inline src/pages/Portfolio/ImpactDashboard/YourDonations.vue:186]: Is the goal to get the last X number of donations? It looks like if the total donations is greater than 100, we get the donations for 100+ transactions.

File: src/pages/Portfolio/ImpactDashboard/YourDonations.vue [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+					}
+				}).finally(() => {
+					this.loadingPromise = null;
+				});
+			}
+		},
+		whenVisible() {
+			this.fetchAsyncData();
+		},
+		getRemainDonations(transactionFilters, offset) {
```
Comment (dyersituations) [inline src/pages/Portfolio/ImpactDashboard/YourDonations.vue:211]: `getRemainingDonations`?

File: src/pages/Portfolio/ImpactDashboard/YourDonations.vue [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	}
+};
+</script>
+
+<style lang="postcss" scoped>
+.stat-placeholder {
+	margin-bottom: 6px;
+	@apply tw-mt-1 tw-h-4.5;
+
+	@screen md {
```
Comment (dyersituations) [inline src/pages/Portfolio/ImpactDashboard/YourDonations.vue:253]: Do these nested `@screen` usages play nicely with postcss? I know we don't have the nesting module, but I wasn't sure if `@screen` counted as nesting here.

File: src/pages/Portfolio/ImpactDashboard/YourDonations.vue (hunk 1 above)
Comment (dyersituations) [reply src/pages/Portfolio/ImpactDashboard/YourDonations.vue:186]: Is this meant to be a slight performance boost?

Comment (emuvente) [reply src/pages/Portfolio/ImpactDashboard/YourDonations.vue:186]: This is just paging, not a performance thing. We need all the donation transactions to be able to count the total donation amount, and the most results we can get back from the collect result in a single query is 100.

File: src/pages/Portfolio/ImpactDashboard/YourDonations.vue [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+					this.averageDonation = donation.format('$0,0[.]00');
+
+					this.totalDonations = data?.my?.transactions?.totalCount ?? 0;
+					this.currentYearDonations = this.filterAndSumDonations(transactions, currentYear);
+					this.lastYearDonations = this.filterAndSumDonations(transactions, currentYear - 1);
+
+					if (this.totalDonations > this.limit) {
+						let offset = this.limit;
+						while (offset < this.totalDonations) {
+							this.getRemainingDonations(transactionFilters, offset)
```
Comment (emuvente) [inline src/pages/Portfolio/ImpactDashboard/YourDonations.vue:189]: The way this works right now, the entire while loop will execute synchronously, kicking off a bunch of parallel requests for the rest of the results, and then complete before any of the results can come in. The `finally()` callback will be called before the rest of those requests complete.

You could maybe build an array of requests based on the totalCount and then return Promise.all(requests)

**Review by emuvente — APPROVED** [review]
Thanks for the updates, looks good!

**Pull request merged**